src/subdomains/core/history/controllers/transaction.controller.ts

```diff
@@ -422,6 +428,80 @@ export class TransactionController {
     });
   }
 
+  @Put(':id/invoice')
+  @ApiBearerAuth()
+  @UseGuards(AuthGuard(), new RoleGuard(UserRole.USER), IpGuard, UserActiveGuard)
+  @ApiOkResponse({ type: InvoiceDto })
+  async generateInvoiceFromTransaction(@GetJwt() jwt: JwtPayload, @Param('id') id: string): Promise<InvoiceDto> {
+    const transaction = await this.transactionService.getTransactionById(+id, {
+      userData: true,
+      buyCrypto: { outputAsset: true, buy: true, cryptoRoute: true, cryptoInput: { asset: true } },
+      buyFiat: { outputAsset: true, sell: true, cryptoInput: { asset: true } },
+      refReward: { user: { userData: true } },
+    });
+
+    if (!transaction) throw new BadRequestException('Transaction not found');
+    if (!transaction.userData.isDataComplete) throw new BadRequestException('User data is not complete');
+    if (transaction.amlCheck && transaction.amlCheck !== CheckStatus.PASS)
+      throw new BadRequestException('AML check not passed');
+    if (transaction.userData.id !== jwt.account) throw new ForbiddenException('Not your transaction');
+
+    const { transactionType, currency } = await this.getInvoiceDetails(transaction);
+
+    return {
+      invoicePdf: await this.swissQrService.createInvoiceFromTx(transactionType, transaction, currency),
+    };
+  }
+
+  private async getInvoiceDetails(
+    transaction: Transaction,
+  ): Promise<{ transactionType: TransactionType; currency: 'CHF' | 'EUR' }> {
+    let transactionType: TransactionType;
```

Comment:
Instead of using the two let variables you could return the values in each if directly 👍